File: src/main/java/top/kmar/mi/api/net/message/player/PlayerMessage.java
package top.kmar.mi.api.net.message.player;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.fml.relauncher.Side;
import top.kmar.mi.api.net.ParseResultEnum;
import top.kmar.mi.api.net.message.IMessageHandle;
import top.kmar.mi.api.net.message.ParseAddition;
import top.kmar.mi.api.utils.MISysInfo;

import javax.annotation.Nonnull;

/**
 * <p>以玩家为凭借的服务端-客户端双向通讯
 * <p>处理端：客户端、服务端
 * @author EmptyDreams
 */
public class PlayerMessage implements IMessageHandle<PlayerAddition, ParseAddition> {
	
	public static PlayerMessage instance() {
		return INSTANCE;
	}
	
	private static final PlayerMessage INSTANCE = new PlayerMessage();
	
	private PlayerMessage() { }
	
	@Override
	public ParseAddition parseOnClient(@Nonnull NBTTagCompound message, ParseAddition result) {
		PlayerAddition addition = PlayerAddition.instance(message.getTag("add"));
		boolean applyResult = PlayerHandleRegistry.apply(
				addition.getKey(), null, message.getTag("data"));
		if (applyResult) return result.setParseResult(ParseResultEnum.SUCCESS);
		MISysInfo.err("没有找到可以处理该信息的Handle");
		return result.setParseResult(ParseResultEnum.EXCEPTION);
	}
	
	@Override
	public ParseAddition parseOnServer(@Nonnull NBTTagCompound message, ParseAddition result) {
		PlayerAddition addition = PlayerAddition.instance(message.getTag("add"));
		boolean applyResult = PlayerHandleRegistry.apply(
				addition.getKey(), result.getServicePlayer(), message.getTag("data"));
		if (applyResult) return result.setParseResult(ParseResultEnum.SUCCESS);
		MISysInfo.err("没有找到可以处理该信息的Handle");
		return result.setParseResult(ParseResultEnum.EXCEPTION);
	}
	
	@Override
	public boolean match(@Nonnull Side side) {
		return true;
	}
	
}

File: src/main/java/top/kmar/mi/api/net/messages/player/PlayerMessage.kt
package top.kmar.mi.api.net.messages.player

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.nbt.NBTBase
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import top.kmar.mi.api.net.CommonMessage
import top.kmar.mi.api.net.NetworkLoader
import top.kmar.mi.api.net.handlers.IAutoNetworkHandler
import top.kmar.mi.api.net.handlers.MessageHandlerRegedit
import top.kmar.mi.api.net.handlers.RetryMessage
import top.kmar.mi.api.regedits.others.AutoLoader
import top.kmar.mi.api.utils.TickHelper
import top.kmar.mi.api.utils.expands.clientPlayer

/**
 * 基于玩家的网络交互
 * @author EmptyDreams
 */
@AutoLoader
object PlayerMessage : IAutoNetworkHandler {

    const val key = "mi:player"

    init {
        MessageHandlerRegedit.registry(key, this)
    }

    /** 发送信息到服务端 */
    @JvmStatic
    @SideOnly(Side.CLIENT)
    fun sendToServer(key: String, data: NBTBase) {
        val message = packing(clientPlayer, key, data)
        NetworkLoader.instance.sendToServer(message)
    }

    /** 发送信息到客户端 */
    @JvmStatic
    fun sendToClient(player: EntityPlayer, key: String, data: NBTBase) {
        val message = packing(player, key, data)
        NetworkLoader.instance.sendTo(message, player as EntityPlayerMP)
    }

    private fun packing(player: EntityPlayer, key: String, data: NBTBase): IMessage {
        val message = NBTTagCompound().apply {
            setString("key", key)
            setTag("data", data)
        }
        return CommonMessage(this.key, message)
    }

    override fun parse(message: NBTBase, ctx: MessageContext): RetryMessage? {
        message as NBTTagCompound
        val player = if (ctx.side.isServer) ctx.serverHandler.player else clientPlayer
        val key = message.getString("key")
        val data = message.getTag("data")
        TickHelper.addAutoTask {
            PlayerHandlerRegedit.apply(key, player, data)
            true
        }
        return null
    }

}